### PR TITLE
Wandering isle quest 29419

### PR DESCRIPTION
**Pull request by aquadeus**
**Changes proposed:**

-  Script quest 29419 blizzlike

**Tests performed:**

Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

- Npcs Amberleaf scamp should walk/run randomly when random movement is set and walk when movetargetedhome is set

### CODE REVIEW

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 1]
```diff
@@ -400,10 +402,206 @@ struct npc_instructor_zhi : public ScriptedAI
...
+        });
+    }
+
+    void WaypointReached(uint32 waypointId, uint32 pathId) override
+    {
+        if (pathId == PATH_MOVE_RUN)
+        {
+            switch (waypointId)
+            {
+                case 0:
```
Comment (mdX7) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:524]: magic numbers

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 2]
```diff
@@ -400,10 +402,206 @@ struct npc_instructor_zhi : public ScriptedAI
...
+                {
+                    Talk(SAY_MIN_DIMWIND_TEXT_3);
+                    me->GetMotionMaster()->MovePath(PATH_MOVE_WALK, false);
+                    break;
+                }
+            }
+        }
+        else if (pathId == PATH_MOVE_WALK)
+        {
+            if (waypointId == 3)
```
Comment (mdX7) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:540]: magic numbers

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 3]
```diff
@@ -400,10 +402,206 @@ struct npc_instructor_zhi : public ScriptedAI
...
+};
+
+// 54130 - Amberleaf Scamp
+struct npc_amberleaf_scamp : public ScriptedAI
+{
+    npc_amberleaf_scamp(Creature* creature) : ScriptedAI(creature) {  }
+
+    void MovementInform(uint32 /*type*/, uint32 id) override
+    {
+        if (id == 0)
```
Comment (mdX7) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:572]: magic numbers

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 4]
```diff
@@ -400,10 +402,206 @@ struct npc_instructor_zhi : public ScriptedAI
...
+        else if (pathId == PATH_MOVE_WALK)
+        {
+            if (waypointId == 3)
+            {
+                me->SetFacingTo(0.575958f);
+                me->DespawnOrUnsummon(2s);
+
+                _scheduler.Schedule(Seconds(1), [this](TaskContext /*task*/)
+                {
+                    Unit* summoner = me->ToTempSummon()->GetSummonerUnit();
```
Comment (mdX7) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:547]: crash when `me->ToTempSummon()` null

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 5]
```diff
@@ -400,10 +402,206 @@ struct npc_instructor_zhi : public ScriptedAI
...
+
+                    summoner->RemoveAurasDueToSpell(SPELL_SUMMON_CART_DRIVER);
```
Comment (mdX7) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:549]: crash when summoner null

**Review by mdX7 — CHANGES_REQUESTED** [review]

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 6]
```diff
@@ -400,10 +402,220 @@ struct npc_instructor_zhi : public ScriptedAI
...
+            if (!player->isDead() && player->GetQuestStatus(QUEST_THE_MISSING_DRIVER) == QUEST_STATUS_INCOMPLETE)
+            {
+                Creature* minDimwind = player->FindNearestCreatureWithOptions(20.0f, { .StringId = "npc_min_dimwind" });
+
+                if (!minDimwind)
+                    return false;
+
+                minDimwind->CastSpell(player, SPELL_FORCE_SUMMON_CART_DRIVER, TRIGGERED_FULL_MASK);
+                player->CastSpell(player, SPELL_SERVERSIDE_KILL_CREDIT, TRIGGERED_FULL_MASK);
+                PhasingHandler::OnConditionChange(player); // blizzard adds phase 630 when kill credit but immediately is removed to add it again when Min Dimwind reaches final waypoint
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:464]: remove blizzard from comment

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 7]
```diff
@@ -400,10 +402,225 @@ struct npc_instructor_zhi : public ScriptedAI
...
+            _scheduler.Schedule(10s, [this](TaskContext /*task*/)
+            {
+                me->GetMotionMaster()->MoveTargetedHome();
+            });
+        }
+    }
+
+    void JustReachedHome() override
+    {
+        me->GetMotionMaster()->MoveIdle();
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:604]: Wouldn't `GetMotionMaster()->InitializeDefault()` be more appropriate here?

**Comment by aquadeus** [comment]
> Conflicting files src/server/scripts/Pandaria/zone_the_wandering_isle.cpp

done

File: src/server/scripts/Pandaria/zone_the_wandering_isle.cpp [hunk 8]
```diff
@@ -680,7 +680,7 @@ class at_min_dimwind_captured : public AreaTriggerScript
                     return false;
 
                 minDimwind->CastSpell(player, SPELL_FORCE_SUMMON_CART_DRIVER, TRIGGERED_FULL_MASK);
-                player->CastSpell(player, SPELL_SERVERSIDE_KILL_CREDIT, TRIGGERED_FULL_MASK);
+                player->CastSpell(player, SPELL_SERVERSIDE_DRIVER_CREDIT, TRIGGERED_FULL_MASK);
```
Comment (Shauren) [inline src/server/scripts/Pandaria/zone_the_wandering_isle.cpp:683]: this shows the PR was untested

**Pull request merged**